docarray/array/array/pushpull/__init__.py

```diff
@@ -0,0 +1,244 @@
+import logging
+from abc import abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, Dict, Iterable, Iterator, List, Optional, Type
+
+from typing_extensions import Protocol
+
+__cache_path__ = Path.home() / '.cache' / 'docarray-v2'
+
+if TYPE_CHECKING:  # pragma: no cover
+    from docarray import BaseDocument, DocumentArray
+
+
+class ConcurrentPushException(Exception):
+    """Exception raised when a concurrent push is detected."""
+
+    pass
+
+
+class PushPullLike(Protocol):
+    @staticmethod
+    def list(namespace: str, show_table: bool) -> List[str]:
+        ...
+
+    @staticmethod
+    def delete(name: str, missing_ok: bool) -> bool:
+        ...
+
+    @staticmethod
+    def push(
+        da: 'DocumentArray',
+        url: str,
+        public: bool,
+        show_progress: bool,
+        branding: Optional[Dict],
+    ) -> Dict:
+        ...
+
+    @staticmethod
+    def push_stream(
+        docs: Iterator['BaseDocument'],
+        url: str,
+        public: bool = True,
+        show_progress: bool = False,
+        branding: Optional[Dict] = None,
+    ) -> Dict:
+        ...
+
+    @staticmethod
+    def pull(
+        cls: Type['DocumentArray'],
+        name: str,
+        show_progress: bool,
+        local_cache: bool,
+    ) -> 'DocumentArray':
+        ...
+
+    @staticmethod
+    def pull_stream(
+        cls: Type['DocumentArray'],
+        name: str,
+        show_progress: bool,
+        local_cache: bool,
+    ) -> Iterator['BaseDocument']:
+        ...
+
+
+class PushPullMixin(Iterable['BaseDocument']):
+    """Mixin class for push/pull functionality."""
+
+    __backends__: Dict[str, PushPullLike] = {}
+    document_type: Type['BaseDocument']
+
+    @abstractmethod
+    def __len__(self) -> int:
+        ...
+
+    @classmethod
+    def get_backend(cls, protocol: str) -> PushPullLike:
+        """
+        Get the backend for the given protocol.
+
+        :param protocol: the protocol to use, e.g. 'jinaai', 'file', 's3'
+        :return: the backend class
+        """
+        if protocol in cls.__backends__:
+            return cls.__backends__[protocol]
+
+        if protocol == 'jinaai':
+            from docarray.array.array.pushpull.jinaai import PushPullJAC
+
+            cls.__backends__[protocol] = PushPullJAC
+            logging.info('Loaded Jina AI Cloud backend')
+        elif protocol == 'file':
+            from docarray.array.array.pushpull.file import PushPullFile
+
+            cls.__backends__[protocol] = PushPullFile
+            logging.info('Loaded Local Filesystem backend')
+        elif protocol == 's3':
+            from docarray.array.array.pushpull.s3 import PushPullS3
+
+            cls.__backends__[protocol] = PushPullS3
+            logging.info('Loaded S3 backend')
```

Comment:
info or debug for the logging ?

Reply:
Now that you've pointed it out, yea this makes more sense as debug